File: week-5/protoseo/11286.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        PriorityQueue<Integer> pq = new PriorityQueue<>((o1, o2) -> {
            if (Math.abs(o1) == Math.abs(o2)) {
                return o1 - o2;
            }
            return Math.abs(o1) - Math.abs(o2);
        });

        StringBuilder sb = new StringBuilder();
        while (n-- > 0) {
            int a = Integer.parseInt(br.readLine());
            if (a == 0) {
                if (pq.isEmpty()) {
                    sb.append(0);
                } else {
                    sb.append(pq.poll());
                }
                sb.append('\n');
                continue;
            }
            pq.add(a);
        }
        System.out.println(sb);
    }
}

File: week-05/devxb/11286.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.abs

val absQ = PriorityQueue<Element>()

fun main() {
    BufferedReader(InputStreamReader(System.`in`))
        .use { reader ->
            val N = reader.readLine().toInt()
            for (i in 1..N) {
                val num = reader.readLine().toInt()
                if (num == 0) {
                    absQ.printAndPoll()
                    continue
                }
                absQ.add(Element(abs(num), num))
            }
        }
}

class Element(val absNum: Int, val ordinaryNum: Int) : Comparable<Element> {

    override fun compareTo(other: Element): Int {
        if (absNum < other.absNum) {
            return -1
        }
        if (absNum > other.absNum) {
            return 1
        }
        if (ordinaryNum < other.ordinaryNum) {
            return -1
        }
        return 1
    }
}

fun PriorityQueue<Element>.printAndPoll() {
    if (this.isEmpty()) {
        println(0)
        return
    }
    println(this.poll().ordinaryNum)
}
